### ISSUE TITLE
[Improvement] Better "Add" Mod button (Mod Browser)

### ISSUE THREAD

**Issue by homemmakako** [ATLauncher/ATLauncher#556]
The only thing that annoys me a bit with the current mod browser is the excess of new windows and clicks. (click add button -> open a new window -> select version -> click install)

I feel like a nice improvement would be to replace the add button with a button+dropdown just like the one used for the play button, where you can click the button to play, but press the right dropdown arrow to show extra options. In this case the button would be renamed from "add" to "install"; clicking the button would just install the latest version and clicking the dropdown would show extra options such as "Install specific version" and "View mod page".

This would make the mod install process a lot more fluid and pleasant, while being (I believe) not insanely complex to implement.

Note: Also, another improvement that would go well with this would be a less instrusive "Installing mod" progress bar other than the current new window, but that a whole other topic and right now I can't really think of a good simple solution that wouldn't involve changing the current interface or removing the progress bar entirely.

**Comment by RyanTheAllmighty** [comment]
Yeah the UX/UI sucks and I hate most of it. A lot of it stems from 7 year old decisions and inflexibility/ease of implementation in Java and Swing that the launcher uses.

While I'm happier with how newer features are done, such as the vanilla packs and packs browser tabs, the rest I'm generally unhappy with.

The whole add mods section for adding mods from CurseForge and Modrinth is definitely the next thing I want to tackle, as that whole area and flow is, as you've mentioned, pretty bad.

**Comment by Rosethorns** [comment]
As per requested from the ATL Discord I'm adding my sugggestion for the edit instance mod browser here.

Suggestion: Please add visual feedback to mods in the mod browser window that are already installed in the current instance.

Problem: Currently while the mod browser window is open you are unable to tell what mods you do/do not have installed in the current instance. This is compounded by the fact that the instance window is locked while the mod browser window is open meaning you cannot scan your list of currently installed mods for a reminder.

Thanks for reading.

**Comment by Rosethorns** [comment]
Suggestion: Offer alternative repository results if a searched mod can't be found on the chosen service.

Problem: If I want to primarily fetch mods from Modrinth but the mod isn't available I have to switch and search on CurseForge instead. Small issue but contributes to search fatigue.

Example: I Choose Modrinth as the repository I want ATL to search. I look for Mekanism but its not found on Modrinth. ATL Offers "We couldn't find the mod on Modrinth but these are the results on CurseForge." 

While my examples are specific, this would obviously work regardless of what repository the user has chosen. This should also be implemented such that other repositories would be searched in succession if other repositories become available in the future.

**Comment by Rosethorns** [comment]
Suggestion: In the Edit Mods window, add mod searching.
Problem: In a long list of mods it can be hard to identify at a glance what you are looking for.

Suggestion: In the Edit Mods window, offer multiple list organizations rather than just alphabetical. Such as "last added" or "User defined" (for clarity, when a user can move an items position in a list up or down manually)
Problem: Being restricted to just alphabetical impacts work flow for some power users.

Suggestion: In the Edit Mods window, more visual differentiation between mod entries such as the mods thumbnail next to the entry.
Problem: As it stands the interface is quite sterile and this can lead to difficulty recognizing items at a glance.

**Comment by CoolCat467** [comment]
Someone might have already posted something about this, but if you search for a mod and there are more than one page of results and you go to different pages, but then search something different with less results, it doesn't automatically reset the page you are viewing to the first one, making it seem like there are no results for mod unless you realize you can switch back pages to the first one.

**Comment by davidmyska** [comment]
> As per requested from the ATL Discord I'm adding my sugggestion for the edit instance mod browser here.
> 
> Suggestion: Please add visual feedback to mods in the mod browser window that are already installed in the current instance.
> 
> Problem: Currently while the mod browser window is open you are unable to tell what mods you do/do not have installed in the current instance. This is compounded by the fact that the instance window is locked while the mod browser window is open meaning you cannot scan your list of currently installed mods for a reminder.
> 
> Thanks for reading.

I agree, this is a critical issue that should have been implemented from the beginning, just simply replace the Add button with "Installed" text.

**Comment by AthinCZ** [comment]
> As per requested from the ATL Discord I'm adding my sugggestion for the edit instance mod browser here.
> 
> Suggestion: Please add visual feedback to mods in the mod browser window that are already installed in the current instance.
> 
> Problem: Currently while the mod browser window is open you are unable to tell what mods you do/do not have installed in the current instance. This is compounded by the fact that the instance window is locked while the mod browser window is open meaning you cannot scan your list of currently installed mods for a reminder.
> 
> Thanks for reading.

I agree, this is a critical issue that should have been implemented from the beginning, just simply replace the Add button with "Installed" text or something like that.

**Comment by Rosethorns** [comment]
Option to keep the instance browser open after an instance has been launched.

Example: For the purposes of pack development it would be nice to be able to still have the "open folder." button available for the instance.

**Comment by fruini** [comment]
> As per requested from the ATL Discord I'm adding my sugggestion for the edit instance mod browser here.
> 
> Suggestion: Please add visual feedback to mods in the mod browser window that are already installed in the current instance.
> 
> Problem: Currently while the mod browser window is open you are unable to tell what mods you do/do not have installed in the current instance. This is compounded by the fact that the instance window is locked while the mod browser window is open meaning you cannot scan your list of currently installed mods for a reminder.
> 
> Thanks for reading.

I would love to see this feature being implemented. In addition to that, I also like to have an option to hide all installed mods in Add Mods window so that it will be less cluttered.

**Comment by lilmayu** [comment]
Extracted from: https://github.com/ATLauncher/ATLauncher/issues/718 for easier tracking

Search bar in "Edit Mods" menu:
- The only real "problem" is the inconvenience of having to search for mods manually in the "edit mods" menu, instead of just filling quick search field and having the mods list filtered. For example, MultiMC has a search field in the "mods" category and it's generally a nice UX.
 - Search field for filtering mod list by mod name (or even version, author, etc.)

**Comment by korg** [comment]
Search suggestion.

Curseforge searching is horrible. I was looking to install the mod "mod menu" but it was so hard to search for because so many mods have the word mod in their name or text. Even if you have a link to the mod, often you cant find the mod in the atlaucher curseforge search.

I had the curseforge website link for the mod: https://www.curseforge.com/minecraft/mc-mods/modmenu
and i tried searching modmenu and mod<space>menu with no luck.

Searching for just menu returned 2 pages of results but modmenu was not in that list.

It seems this is a Fabric thing, because if i create a minecraft with fabric instance i can find it. It is still very very confusing though.

Suggestions for curseforge searching:

  * if i search for something, also directly check to see if there is a mod of that name and if it does exist put it first. Eg, looking for modmenu (the curseforge name) looks for a mod at the above url

  * accept curseforge url's in the search bar, so if the user has the curseforge mod page, they can specify *exactly* the mod they are looking for.

  * if you find an exact match and there is some reason that it wont install (Fabric in this case) then maybe still include it but put a note on why it wont install.

I feel your pain for the api you have from curseforge though. ;-/

**Comment by Poopooracoocoo** [comment]
Isn't that an issue on CurseForge's side and not ATL's?

**Comment by lilmayu** [comment]
> Search suggestion.
> 
> Curseforge searching is horrible. I was looking to install the mod "mod menu" but it was so hard to search for because so many mods have the word mod in their name or text. Even if you have a link to the mod, often you cant find the mod in the atlaucher curseforge search.
> 
> I had the curseforge website link for the mod: https://www.curseforge.com/minecraft/mc-mods/modmenu
> and i tried searching modmenu and mod<space>menu with no luck.
> 
> Searching for just menu returned 2 pages of results but modmenu was not in that list.
> 
> It seems this is a Fabric thing, because if i create a minecraft with fabric instance i can find it. It is still very very confusing though.
> 
> Suggestions for curseforge searching:
> 
>   * if i search for something, also directly check to see if there is a mod of that name and if it does exist put it first. Eg, looking for modmenu (the curseforge name) looks for a mod at the above url
> 
>   * accept curseforge url's in the search bar, so if the user has the curseforge mod page, they can specify *exactly* the mod they are looking for.
> 
>   * if you find an exact match and there is some reason that it wont install (Fabric in this case) then maybe still include it but put a note on why it wont install.
> 
> I feel your pain for the api you have from curseforge though. ;-/
> 
> 

Having the ability to paste the link there or see incompatible mods would be great. Incompatible mod would have a slightly red background for example. 

But I think they want to rework the download menu little bit, so we will see.

**Comment by CoolCat467** [comment]
A better solution than link in the search bar would be a dedicated menu for add mod from link or something, because otherwise every time someone searches something the system would have to deceiver if the search query is a link to a website or just something to search

**Comment by RyanTheAllmighty** [comment]
Just to update on this. Work has been going on for a couple weeks now on this uplift work. So far going well, I expect it to take me at least another month given how much has been done vs what's left and how much time I'll likely have to go with.

Trying to address as many concerns as I can from this thread, I think it'll be miles better than what is there now, even if not perfect.

So yeah going to aim for release end of May or thereabouts.

**Comment by RyanTheAllmighty** [comment]
Remaining tasks on branch feature/GH-556-new-edit-instance-screen:

- [ ] Refactor to use ViewModels
- [ ] Refactor to use RXJava
- [ ] Issues with Java's pass by value means when an instance is updated or reinstalled, the instance object passed to each section/edit instance dialog is now out of date with the actual instance data
- [ ] Would like there to be a message that shows up when changing Minecraft version for a vanilla instance to bring up the reinstall prompt for all mods
- [ ] Mod browser is still using the same old one, that needs a bit of a spit shine (might break this one out, not sure yet)
- [ ] Since making the edit instance dialogs become non blocking to the main window, there's nothing stopping someone from reinstalling an instance in both the edit instance window and the main window

**Comment by CorneliusCornbread** [comment]
You should also add the ability to install and manage shaderpacks as well. Prism already supports shaderpack management

**Comment by RyanTheAllmighty** [comment]
We already do that with the current version

**Comment by CorneliusCornbread** [comment]
> We already do that with the current version

oh I see, _only_ modrinth has support for shaders. I was looking at CurseForge, it seems it's missing from the dropdown for CurseForge

**Comment by RyanTheAllmighty** [comment]
Yeah unfortunately CurseForge do not provide a category for Shader Packs. Shader Packs on CurseForge are just a search query of "Shaders" in the customisation category.

I've looked at this in the past and means you cannot search for Shader Packs on CurseForge reliably. It's not really usable the same way Modrinth is unfortunately
